backend/src/main/controller/PromotionController.ts

```diff
@@ -0,0 +1,188 @@
+  PromotionQueryDTO,
+  PromotionQueryValidation,
+} from '../validation/PromotionQueryValidation';
+import { IdValidation } from '../validation/IdValidation';
+
+/* eslint-disable  @typescript-eslint/no-explicit-any */
+export class PromotionController {
+  /**
+   * Retrieves all promotions and their discounts
+   * * First we need to validate the query params and cast that into a PromotionQueryDTO
+   * * Then we apply the query options into the query builder depending on which properties are present
+   * * Now we execute the query builder and return its results
+   */
+  getAllPromotions = async (
+    request: Request,
+    response: Response,
+    next: NextFunction
+  ): Promise<any> => {
+    try {
+      let queryBuilder = getCustomRepository(PromotionRepository)
+        .createQueryBuilder('promotion')
+        .innerJoinAndSelect('promotion.discount', 'discount');
+
+      if (request.query) {
+        const promotionQuery: PromotionQueryDTO = await PromotionQueryValidation.schema.validateAsync(
+          request.query,
+          { abortEarly: false }
+        );
+        queryBuilder = this.applyQueryOptions(queryBuilder, promotionQuery);
+      }
+
+      const promotions = await queryBuilder.getMany();
+      return response.send(promotions);
+    } catch (e) {
+      return next(e);
+    }
+  };
+
+  /**
+   * Depending on which properties are defined inside promotionQuery, we add those properties into our query for the queryBuilder to execute.
+   */
+  private applyQueryOptions(
+    queryBuilder: SelectQueryBuilder<Promotion>,
+    promotionQuery: PromotionQueryDTO
+  ): SelectQueryBuilder<Promotion> {
+    if (promotionQuery?.category) {
+      queryBuilder = queryBuilder.andWhere('promotion.category = :category', {
+        category: promotionQuery.category,
+      });
+    }
+
+    if (promotionQuery?.cuisine) {
+      queryBuilder = queryBuilder.andWhere('promotion.cuisine = :cuisine', {
+        cuisine: promotionQuery.cuisine,
+      });
+    }
+
+    if (promotionQuery?.discountType) {
+      queryBuilder = queryBuilder.andWhere('discount.type = :type', {
+        type: promotionQuery.discountType,
+      });
+    }
+
+    if (promotionQuery?.expirationDate) {
+      queryBuilder = queryBuilder.andWhere(
+        'promotion.expirationDate >= :date',
+        {
+          date: new Date(promotionQuery.expirationDate),
+        }
+      );
+    }
+
+    if (promotionQuery?.name) {
+      let name = promotionQuery.name;
+
+      // first escape % and _ (request needs to show it as https://stackoverflow.com/questions/17342671/pass-a-percent-sign-in-a-url-and-get-exact-value-of-it-using-php)
+      // this way users can query promotions that contain % or _
+      name = name.replace('%', '\\%');
+      name = name.replace('_', '\\_');
+
+      // add wildcards
+      name = `%${name.replace(/ /g, '%')}%`;
+
+      queryBuilder = queryBuilder.andWhere('promotion.name ilike :name', {
+        name: name,
+      });
```

Comment:
So this is how I'm thinking we could the FE submits queries for promotion names.

Say for example the user wants to find promotions that contain "promo hi". We would then add wildcards `%` at the start and end of the string, and in spaces. That way we can leverage `postgres`'s `ILIKE` comparator.

In cases where the user is searching for `50%`, then we don't want the % to act as a wildcard. THis will be the responsibility of the FE to encode % into `%25` (see the link in the code). That way we can escape % by turning it into `\\%`, then adding the wildcards.

@na-wu would there be a better way for this search query?

Reply:
Edit: disregard this, I tried to implement using postgres's full text search. @na-wu lmk your thoughts if you get the chance to take a look! It's mainly this commit https://github.com/ubclaunchpad/foodies/commit/0f1728b9c40010d69457f95babf05b63100f1524